src/metricscollector/cf/cf_client_test.go

```diff
@@ -228,4 +241,163 @@ var _ = Describe("CfClient", func() {
 			})
 		})
 	})
+
+	Describe("RefreshAuthToken", func() {
+		BeforeEach(func() {
+			cfc = NewCfClient(&conf.Cf, lager.NewLogger("cf"))
+		})
+
+		JustBeforeEach(func() {
+			authToken, err = cfc.RefreshAuthToken()
+		})
+
+		Context("when not logged in", func() {
+			BeforeEach(func() {
+				fakeCC.AppendHandlers(
+					ghttp.CombineHandlers(
+						ghttp.VerifyRequest("GET", PathCfInfo),
+						ghttp.RespondWith(200, bytes.Replace(infoBody, []byte("test-oauth-endpoint"), []byte(fakeLoginServer.URL()), -1)),
+					),
+				)
+			})
+
+			Context("when login server returns a non-200 status code for login", func() {
+				BeforeEach(func() {
+					fakeLoginServer.AppendHandlers(
+						ghttp.CombineHandlers(
+							ghttp.VerifyRequest("POST", PathCfAuth),
+							ghttp.RespondWith(401, ""),
+						),
+					)
+				})
+
+				It("should error", func() {
+					Expect(err).To(MatchError(MatchRegexp("request token grant failed: .*")))
+				})
+
+			})
+
+			Context("when login server returns a 200 status code for login", func() {
+				BeforeEach(func() {
+					fakeLoginServer.AppendHandlers(
+						ghttp.CombineHandlers(
+							ghttp.VerifyRequest("POST", PathCfAuth),
+							ghttp.RespondWith(200, authBody),
+						),
+					)
+				})
+
+				It("logs in and returns valid token", func() {
+					Expect(err).NotTo(HaveOccurred())
+					Expect(authToken).To(Equal("test-access-token"))
+					Expect(cfc.GetTokens().AccessToken).To(Equal("test-access-token"))
+					Expect(cfc.GetTokens().RefreshToken).To(Equal("test-refresh-token"))
+					Expect(cfc.GetTokens().ExpiresIn).To(Equal(int64(12000)))
+				})
+
+			})
+
+		})
+
+		Context("when already logged in", func() {
+			BeforeEach(func() {
```

Comment:
The setup is wrong.
If we are already logged in, we should not see any request made to PathCFInfo.
It is always best to structure the good path first and then do all the error cases.